flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+To get familiar with FileCheck tests, it is worth having a look at `https://llvm.org/docs/CommandGuide/FileCheck.html`.
+A few test cases must be added in `intrinsic-procedures.f90` to test that the fir output looks like what is expected.
+To do quick eye checking during development, the FIR produced for a Fortran file can be dumped with `bbc -emit-fir`.
+
+### Adding intrinsics in the FIR and Codegen
+So far, nothing to do there. No FIR operation implements Fortran intrinsic procedures.
+
+## Tips
+
+Be sure to respect alphabetical order when adding a new intrinsic in one of the tables or
+if statements. They are usually sorted according to intrinsic names.
+
+## An End-to-End example: TRIM
+TRIM is a transformational intrinsic operating on characters.
+It takes a scalar character and returns a scalar character. See Fortran standard for more details about it.
+### TRIM name resolution and semantic checks
+In lib/Evaluate/intrinsics.cpp, the table contains a line:
+```
+ {"trim", {{"string", SameChar, Rank::scalar}}, SameChar, Rank::scalar, IntrinsicClass::transformationalFunction}
+```
+If defines a transformational intrinsic named "trim" that takes a scalar character argument
+named "string" and returns a scalar character of the same kind as the argument.
+There is nothing more needed for name resolution and semantic checks.
+
+### TRIM front-end folding
+Trim can appear in constant expressions, as per Fortran 2018 section 10.1.12 point 1 sub-point (6).
+
+In `lib/Evaluate/fold-character.cpp`, there is case to fold trim that looks like:
+```
+  } else if (name == "trim") { // not elemental
+    if (auto scalar{
+            GetScalarConstantArguments<T>(context, funcRef.arguments())}) {
+      return Expr<T>{Constant<T>{
+          CharacterUtils<KIND>::TRIM(std::get<Scalar<T>>(*scalar))}};
+    }
+  }
+```
+It uses a helper function operating on the `Scalar` abstract representation. `Scalar<T>` when T is a character kind 1
+are simply `std::string`, so `CharacterUtils<1>::TRIM` actually simply implements Trim on a `std::string`.
+The folding code above mainly try to extract a constant from the argument, and if successful, repackage the result
+of the helper in a new typed expression representation (`Expr<T>`).
+### TRIM runtime
+Trim runtime API is defined in `runtime/character.h` as:
+
+```
+void RTNAME(Trim)(Descriptor &result, const Descriptor &string,
+    const char *sourceFile = nullptr, int sourceLine = 0);
+```
+
+The `RTNAME` macro adds some mangling to the runtime to avoid collision with user symbols and handle
+runtime library versioning. Here, it only matters to know it must be used in all function declaration of
+the runtime API.
+
+Trim runtime takes a descriptor for the result that it will allocate, and a constant descriptor for the string argument.
+On top of that, it takes optional source file and line arguments, that are here to print source location if the runtime
```

Comment:
Optional: the default parameters in the header files are parameters that a C++ compiler would supply at the call site. They are meaningless in terms of our code generator. Said another way, we are writing the compiler, so we have to supply these values regardless. No free lunch.

Reply:
Added:
> Note that C++ runtime optional arguments must still be created explicitly in lowering. They are only indications of what the runtime expects as default values. Here, if no source 
 location were available, lowering code would have to generate `nullptr` and `0` and pass them explicitly in FIR.